src/panels/config/devices/ha-devices-data-table.ts

```diff
@@ -0,0 +1,281 @@
+      narrow
+        ? {
+            device: {
+              title: "Device",
+              sortable: true,
+              filterKey: "name",
+              filterable: true,
+              direction: "asc",
+              template: (device: DeviceRowData) => {
+                const battery = device.battery_entity
+                  ? this.hass.states[device.battery_entity]
+                  : undefined;
+                // Have to work on a nice layout for mobile
+                return html`
+                  ${device.name_by_user || device.name}<br />
+                  ${device.area} | ${device.integration}<br />
+                  ${battery
+                    ? html`
+                        ${battery.state}%
+                        <ha-state-icon
+                          .hass=${this.hass!}
+                          .stateObj=${battery}
+                        ></ha-state-icon>
+                      `
+                    : ""}
+                `;
+              },
+            },
+          }
+        : {
+            device_name: {
+              title: "Device",
+              sortable: true,
+              filterable: true,
+              direction: "asc",
+            },
+            manufacturer: {
+              title: "Manufacturer",
+              sortable: true,
+              filterable: true,
+            },
+            model: {
+              title: "Model",
+              sortable: true,
+              filterable: true,
+            },
+            area: {
+              title: "Area",
+              sortable: true,
+              filterable: true,
+            },
+            integration: {
+              title: "Integration",
+              sortable: true,
+              filterable: true,
+            },
+            battery: {
+              title: "Battery",
+              sortable: true,
+              type: "numeric",
+              template: (batteryEntity: string) => {
+                const battery = batteryEntity
+                  ? this.hass.states[batteryEntity]
+                  : undefined;
+                return battery
+                  ? html`
+                      ${battery.state}%
+                      <ha-state-icon
+                        .hass=${this.hass!}
+                        .stateObj=${battery}
+                      ></ha-state-icon>
+                    `
+                  : html`
+                      -
+                    `;
+              },
+            },
+          }
+  );
+
+  protected render(): TemplateResult {
+    return html`
+      <ha-data-table
+        .columns=${this._columns(this.narrow)}
+        .data=${this._devices(
+          this.devices,
+          this.entries,
+          this.entities,
+          this.areas,
+          this.domain,
+          this.hass.localize
+        ).map((device: DeviceRowData) => {
+          // We don't need a lot of this data for mobile view, but kept it for filtering...
+          const data: DataTabelRowData = {
```

Comment:
Why shouldn't we just return `device` ? It's the exact same keys except for `battery_entity`, which we could change. Then we don't have to create a new object.